### ISSUE TITLE
[Feature] Implement array-access syntax `IDENTIFIER[INDEX]` for strings

### ISSUE THREAD

**Issue by Luna-Klatzer** [Kipper-Lang/Kipper#372]
### Is there an existing proposal for this?

- [X] I have searched the existing issues

### This feature does not exist in the latest version

- [X] I am using the latest version

### Proposal

Implement new `IDENTIFIER[INDEX]` syntax, which allows indices to be used to access substrings inside a string. This should behave similarly to arrays, where the index points to a value inside the string.

For example:

- Using a constant string:
  ```ts
  "a string"[2] // 's' 
  ```

- Using a variable:
  ```ts
  var x: string = "a string";
  x[2] // 's' 
  ```

This behaviour though should be implemented in a different way than JavaScript, since usual JavaScript indices only allow positive integers, but especially for slices, additional syntax sugar like from Python would be helpful to simplify the array access.

For example, the following indices should also be usable:

- Negative indexes to start from the end of an array:
  ```ts
   "0123456789"[-2] // '8' 
  ```

- Indice ranges with the same behaviour as `String.prototype.slice` (Start is inclusive, end is exclusive. Additionally, if end is greater than str.length then it will return everything until the end)
  ```ts
  "0123456789"[0:3] // '012'

  "0123456789"[2:5] // '234'

  "0123456789"[8:15] // '89'
  ```

### Exact behaviour / changes you want

- Implement new indices syntax for strings, additionally with range indices.
- Implement semantic checking for accessing and throw errors if the indices are NOT of type `num`.
- Add a new runtime function called `index(arrayLike: str, start: num, end: num | undefined)` (which for now only supports strings), which implements the behaviour for the string indices.

**Comment by Luna-Klatzer** [comment]
Possible function implementation of `index`:

```js
__kipper.index = (arrayLike, start, end) => {
  if (!arrayLike || !start) return arrayLike;
  if (end) {
    return arrayLike.slice(start, end)
  }
  return start >= 0 ? arrayLike[start] : arrayLike.charAt(arrayLike.length + start)
}
```

This is just a proof of concept though.

**Comment by Luna-Klatzer** [comment]
Will be implemented alongside #67, as it already requires an implementation for member access syntax

**Comment by Luna-Klatzer** [comment]
This issue required some reworking in the internal structure of the Antlr4 Parser and the AST generation process, but with some tweaks and custom implementations, the generation works like before.

More info in the following commits:
- 81d6c4a93862e90d18cf1c2ce65b927cfdbb7c9e
- 40aa0a0b4d46363d2d8361f26ea4d8f910ecbbf2
- fb4c8ecfe987799e2d290014cbdceaa6ab91c5f1
- a2d6d175dbe756e0aa605944f4ab3a463abbe014

**Comment by Luna-Klatzer** [comment]
> Will be implemented alongside #67, as it already requires an implementation for member access syntax

Now with the implementation of the bracket access/indexing functionality and syntax for strings, dot notation was also alongside it mostly implemented, but core semantic checking and type checking will be moved to `v0.11.0`, as objects are not planned for `v0.10.0` . As such the issue will be now postponed.

**Comment by Luna-Klatzer** [comment]
Implemented by #402 and will be released in `v0.10.0-alpha.7`
